servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/FilterableClientToClient.java

```diff
@@ -196,4 +203,65 @@ public Completable closeAsyncGracefully() {
+                                                                final Object lock) {
+        return Single.defer(() -> {
+            final ContextMap context = request.context();
+
+            if (SKIP_CONCURRENT_REQUEST_CHECK) {
+                return executeRequest(requester, request, context, strategy).shareContextOnSubscribe();
+            }
+
+            // Prevent concurrent execution of the same request through the same layer.
+            // In general, we do not expect users to execute the same mutable request concurrently. Therefore, the cost
+            // of synchronized block should be negligible for most requests, unless they messed up with reactive streams
+            // chain and accidentally subscribed to the same request concurrently. This protection helps them avoid
+            // ambiguous runtime behavior caused by a corrupted mutable request state.
+            final Object inFlight;
+            // Note that because request.context() lazily allocates a new ContextMap, there is a risk that
+            // synchronization will happen on two different contexts. However, this is acceptable compromise because:
+            //  - Most likely users subscribe 2+ times to the same request from the same thread.
+            //  - This is the best effort protection, giving users at least one rejection should be enough to let them
+            //    know their code is incorrect and should be rewritten.
+            synchronized (context) {
+                // We do not override lock because other layers may already set their own one.
+                inFlight = context.putIfAbsent(HTTP_IN_FLIGHT_REQUEST, lock);
+            }
+            if (lock.equals(inFlight)) {
+                return Single.<StreamingHttpResponse>failed(new RejectedSubscribeException(
+                        "Concurrent execution is detected for the same mutable request. Only a single execution is " +
+                                "allowed at any point of time. Otherwise, request data structures can be corrupted. " +
+                                "To avoid this error, create a new request for every execution or wrap every request " +
+                                "creation with Single.defer() operator."))
+                        .shareContextOnSubscribe();
+            }
+
+            Single<StreamingHttpResponse> response = executeRequest(requester, request, context, strategy);
+            if (inFlight == null) {
+                // Remove only if we are the one who set the lock.
+                response = response.beforeFinally(() -> {
+                    synchronized (context) {
+                        context.remove(HTTP_IN_FLIGHT_REQUEST);
+                    }
+                });
+            }
+            return response.shareContextOnSubscribe();
+        });
+    }
+
+    private static Single<StreamingHttpResponse> executeRequest(final StreamingHttpRequester requester,
```

Comment:
Can we change the method name? This is so close in signature to the other that it might be beneficial to have different names.

Reply:
Changed to `setStrategyAndExecute`, open for a better suggestions

Reply:
Nearly anything different will work. I was just lost for a moment because it looked like a recursive call before I examined the parameter types. 😂